src/relayserver/RelayServer.ts

```diff
@@ -429,31 +435,40 @@ export class RelayServer extends EventEmitter {
 
   async replenishWorker (workerIndex: number): Promise<void> {
     const workerAddress = this.getAddress(workerIndex)
-    const workerBalance = toBN(await this.contractInteractor.getBalance(workerAddress))
+    const workerBalance = await this.getWorkerBalance(workerIndex)
     if (workerBalance.lt(toBN(this.workerMinBalance))) {
       const refill = toBN(this.workerTargetBalance).sub(workerBalance)
       const balance = await this.getManagerBalance()
+      const managerHubBalance = await this.relayHubContract?.balanceOf(this.managerAddress) ?? toBN(0)
       console.log(
         `== replenishWorker(${workerIndex}): mgr balance=${balance.div(
-          toBN(1e18)).toString()} worker balance=${workerBalance.div(
+          toBN(1e18)).toString()}  manager hub balance=${managerHubBalance.div(toBN(1e18)).toString()} 
+          worker balance=${workerBalance.div(
           toBN(1e18)).toString()} refill=${refill.div(toBN(1e18)).toString()}`)
-      if (refill.lt(balance.sub(toBN(minimumRelayBalance)))) {
+      if (refill.lt(managerHubBalance.sub(toBN(minimumRelayBalance)))) {
+        const method = this.relayHubContract?.contract.methods.withdraw(toHex(refill), workerAddress)
+        await this._sendTransaction({
+          signerIndex: 0,
+          destination: this.relayHubContract?.address as string,
+          method
+        })
+      } else if (refill.lt(balance.sub(toBN(minimumRelayBalance)))) {
         await this._sendTransaction({
           signerIndex: 0,
           destination: workerAddress,
           value: toHex(refill),
           gasLimit: mintxgascost.toString()
         })
-        await this.getManagerBalance()
       } else {
-        console.log(
-          `== replenishWorker: can't replenish: mgr balance too low ${balance.div(toBN(1e18)).toString()} refill=${refill.div(
-            toBN(1e18)).toString()}`)
+        const message = `== replenishWorker: can't replenish: mgr balance too low ${balance.div(toBN(1e18)).toString()} refill=${refill.div(
+          toBN(1e18)).toString()}`
+        this.emit('fundingNeeded', message)
```

Comment:
To the point of our conversation earlier today: I believe here is an example of an 'exception' you might have caught and know exactly what you want to say to the higher-tier process - you need more money. It is kind of an exception, but a well anticipated one - money go fast. This can be emitted as an event, but is there another way?
What if what you did was calling a purpose-built callback function?

```this.callbacks.onFundingNeeded()```? You do not interrupt the execution flow, do not lose the stack trace or step in functionality, and you know for sure that this will be handled. 

HttpServer's behaviour on 'fundingNeeded' can easily be tested or mocked, and the API for the method can be easily declared in TypeScript, like:
```function onFundingNeeded(message: string, address: Address )```

Reply:
In my view, event-driven approach is like if postmen threw all mail of a street to a single trash can, so that anyone expecting mail would be able to sort through it to find his and ignore the rest of mail and trash. And in case there is only a single recipient for messages, it becomes even less clear why not use a post box?
Sorry for bad metaphores

Reply:
"You do not interrupt the execution flow, do not lose the stack trace or step in functionality, and you know for sure that this will be handled." - I don't understand what you mean here, but this is certainly not the place to do it.